---
title: "Batch effect correction"
author: "Andrea Villanueva Raisman"
output: html_document
date: "2023-11-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r, packages, echo = FALSE}
library(tidyverse)
library(readxl)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(modelr)
library(devtools)
library(HarmonizR)
library(gridExtra)
```


```{r, import_data, echo = FALSE}

#Read in data from Skyline's report
data <- read_csv2('Peptide Ratio Results.csv')

#Get rid of QTags
data <- subset(data, Protein != "QTag")

#Transform from character to numeric class
data$`Ratio To Standard` <- as.numeric(sub(",", ".", data$`Ratio To Standard`))
data$`DotProductLightToHeavy` <- as.numeric(sub(",", ".", data$`DotProductLightToHeavy`))

#Separation of plate (batch) and wells
data <- separate(data, `Replicate Name`, into = c("Plate", "Well"), sep = "_")
data$Plate <- as.numeric(sub(",", ".", data$`Plate`))

# Extract last three characters from 'Well'
data$Well <- substr(data$Well, nchar(data$Well)-2, nchar(data$Well))

#Filter peptides with low dotp
filtered_data <- data %>%
  filter(`DotProductLightToHeavy` > 0.85)
```

# Introduction: batch effect

It is common for proteomic datasets to show high variability, some of which is due to technical noise and bias and not to actual biological variability. Similarly, proteomic datasets can have high numbers of missing values. Even in a singular study, differences between batches might emerge from sample preparation (e.g., laboratory technician, conditions of the equipment, digestion efficiency, etc.) and the variability inherent to the performance of certain analytical tools such as mass spectrometers (Leek et al., 2010). Integrating data from disparate experimental setups requires addressing technical variations, often referred to as batch effects (Voss et al., 2022). 

Batch effects can hinder the extraction of the biological signal and potentially generate artifacts (Čuklina et al., 2019). It is frequent for systematic bias to be an issue in studies in which only a limited number of samples can be prepared or profiled at a time. To impact of batch effects can be lessened through statistical approaches known as batch effect correction or reduction. This can be both prior to the experiments during the experimental design stage (e.g., adding internal standards) and after the serial measurements have been performed during data processing (e.g., through sophisticated computational strategies).

In this study, samples were distributed into 33 different 96-well plates. Each plate had 88 samples in 11 columns and the last column was left empty until the moment of sample digestion, when in-house plasma was added to the top three wells in the column. The objective was to use these as experimental controls. Thus, a first approach to batch effect analysis is looking at the mass spectrometry results for the in-house plasma across all plates. This analysis is based on the ratios of light (endogenous) peptides to the heavy standards. As mentioned above, technical variations due to instrumental drift, sample preparation or instrument sensitivity over time are common in mass spectrometry. In this context, normalizing to an internal standard can enable better comparison between samples by accounting for these fluctuations, ensuring that differences observed are more likely due to biological differences rather than technical artifacts. 

# HarmonizR

While batch effect reduction strategies have been well established for omics such as transcriptomics, other fields with more emerging technologies face a different landscape. A common problem is the dependence on data imputation (Voss et al., 2020). Data imputation is the process of estimating missing or incomplete data values in a dataset. It can be a crucial step in data pre-processing to enable the use of packages or tools that with no tolerance for missing data. However, imputation techniques aim to fill in missing values with estimated or predicted values based on the available information in the dataset and this is prone to error.

Missing data leads sample size reduction and the potential exclusion of incomplete cases may lead to reduced precision or efficiency in statistical analysis, especially if the missing data is not completely random (MAR). However, avoiding imputation has the advantage of retaining the original structure and integrity of the dataset, preventing potential distortions introduced by imputed values. Furthermore, imputation methods can introduce bias, especially if the missing data is not missing at random (MNAR). On the other hand, working with available data only makes the analysis more straightforward without the need for complex imputation strategies, saving time and computational resources and rendering more transparent results which reflect the actual observed values, making it easier to interpret and explain.

For this reason, the batch effect reduction is here performed using the harmonization tool HarmonizR, which can handle missing data without data imputation techniques (Voss et al., 2020). To do so, it capitalizes on the inherent structure of the available data, leveraging a method involving matrix dissection. In practical terms, this involves analyzing the existing structure of the dataset to identify commonalities and inherent patterns that allow for effective alignment, standardization and reduction of technical variations across batches. In the same line, matrix dissection is a technique used in data analysis to break down a matrix into its components. In the context of data harmonization, it involves extracting essential information or patterns from a dataset organized in a matrix format. In this way, HarmonizR minimizes the loss of data by using the available data structure efficiently. 

HarmonizR implements two batch effect reduction methods: ComBat and limma's removeBatchEffect() function. ComBat is widely used for correcting batch effects in high-dimensional data, particularly in genomics. It uses empirical Bayes frameworks to adjust for batch effects while preserving biological variation, using observed data to estimate parameters in a statistical model. limma (linear models for microarray data) is an R package that uses a linear batch correction method primarily for the analysis of microarray data. The removeBatchEffect() function in limma is designed to remove batch effects from expression data, allowing users to adjust for technical variation between batches or groups.

HarmonizR's principle is (Voss et al., 2022):
    1. Combining previously processed datasets (e.g., low quality data points have been excluded) from different experiments into a matrix with all samples and proteins - or peptides - detected in at least one batch
    2. Dissecting a missing value-dependent matrix to enable batch effect correction on sub-matrices. Initially, HarmonizR scans the input matrix for missing values. If a batch has less than 2 values for the particular protein or peptide, the batch itself will be considered as missing.
    3. The chosen batch effect reduction method is performed on each sub-data frame. However, only proteins present in more than one batch will be harmonized.
    4. The user can choose between limma’s removeBatchEffect() function or ComBat based on the distribution of the data. ComBat allows parametric and non-parametric correction with or without scale adjustment. 
    5. The batch-corrected sub-matrices are merged back into a harmonized matrix and those proteins or peptides only present in one batch are added to this matrix.

There are two compulsory arguments for the harmonizR function: raw data and description file which contains the batch indications. However, there are four other optional arguments which include the chosen algorithm (ComBat by default or limma), ComBat_mode (from 1 to 4 and only if ComBat is the chosen algorithm), plot (which can be set to "samplemeans", "featuremeans" or "CV") and output file (a .tsv file) (Schlumbohm, 2022). ComBat modes are based on two parameters, which are par.prior (TRUE for parametric estimations and FALSE for non-parametric ones) and mean.only (TRUE if adjusting the means but not the individual batch effects and FALSE if adjusting both). Mode 1 (selected by default), sets par.prior=TRUE and mean.only=FALSE, mode 2 sets par.prior=TRUE and mean.only=TRUE, mode 3 sets par.prior=FALSE and mean.only=FALSE and mode 4 sets par.prior=FALSE and mean.only=TRUE. 

Using the ComBat algorithm in HarmonizR enables a higher degree of flexibility to choose parameters. By setting mean.only to FALSE, the adjustment of both mean and variance (batch effects) can be ensured. On the other hand, from the QC analysis and filtering of the data, certain assumptions can be relied upon to indicate that using parametric estimations is appropriate. Overall, the mass spectrometry measurements follow a normal (Gaussian) distribution and extreme outliers have been filtered out in each plate. Thus, the harmonizR function is applied with its default parameters. However, this will be explored in three instances: using the filtered raw data and considering each plate as a batch, using the same data but considering four groups based on the visual observation of APOA1 and VTCN and median centering the data before harmonizing using each plate as a batch.

Median centering is a data preprocessing technique commonly used in statistics because it can:
    1. Remove the overall shift or bias: by  subtracting the median value of a variable (here, peptide per plate) from each data point in that variable, the center of the data is shifted to zero without altering the shape of the distribution. This process makes the analysis less sensitive to extreme values or outliers.
    2. Facilitating interpretability and comparison: it allows for a clearer understanding of how individual observations or groups deviate from the central tendency represented by the median.
    3. Normalization: while not strictly a normalization technique, it assists in the process of putting variables on a similar scale. It can thus be useful before performing  batch correction.


```{r, correction_by_plate, echo=FALSE}
# Selecting specific columns to create a new data frame
reduced_df <- filtered_data[c("Plate","Well", "Protein", "Peptide", "Ratio To Standard")]

# Pivoting the data frame from long to wide format
reduced_df <- reduced_df %>%
group_by(Plate, Well, Protein, Peptide) %>%
  summarise(`Ratio To Standard` = mean(`Ratio To Standard`))

wide_df_temp <- pivot_wider(reduced_df, names_from = c(Plate, Well), values_from = `Ratio To Standard`)
wide_df <- data.frame(wide_df_temp[, -(1:2), drop = FALSE])
col_names <- c(colnames(wide_df_temp))
colnames(wide_df) <- c(colnames(wide_df_temp))[-(1:2)]
row_names <- paste(wide_df_temp$Protein, wide_df_temp$Peptide, sep = "-")
rownames(wide_df) <- row_names

# Making a description dataframe where we take each well as a batch
batch_temp <- data.frame(colnames(wide_df))
batch <- separate(batch_temp, colnames.wide_df., into = c("Plate", "Well"), sep = "_")
batch <- c(as.integer(batch$Plate))

des_df <- data.frame(ID = colnames(wide_df), sample = 1:ncol(wide_df), batch = batch)

invisible(harmonizR(wide_df, des_df, output_file = "in-house_batch_by_plate"))
```

```{r, echo=FALSE}
#Save data to variable
cured_data_temp <- read_tsv('in-house_batch_by_plate.tsv')
```

```{r, echo=FALSE}
# Using pivot_longer to transform to long format
long_cured_data <- cured_data_temp %>%
  pivot_longer(cols = -1, names_to = "Replicate Name", values_to = "Ratio To Standard")

#Separation of plate (batch) and wells
batch_by_plate <- separate(long_cured_data, `...1`, into = c("Protein", "Peptide"), sep = "-") %>%
  separate(`Replicate Name`, into = c("Plate", "Well"), sep = "_")

batch_by_plate$Plate <- as.numeric(sub(",", ".", batch_by_plate$`Plate`))

# Extract last three characters from 'Well'
batch_by_plate$Well <- substr(batch_by_plate$Well, nchar(batch_by_plate$Well) - 2, nchar(batch_by_plate$Well))
```


```{r, correction_by_group, echo=FALSE}
# Making a description dataframe where we take each well as a batch
batch_temp <- data.frame(colnames(wide_df))
batch <- separate(batch_temp, colnames.wide_df., into = c("Plate", "Well"), sep = "_")
batch <- ifelse(batch$Plate %in% 1:6, 1,
                ifelse(batch$Plate == 7, 2,
                       ifelse(batch$Plate %in% 8:30, 3,
                              ifelse(batch$Plate %in% 31:33, 4, NA)
                       )
                )
)

# Create the description data frame 'des_df' based on batch assignment
des_df <- data.frame(ID = colnames(wide_df), sample = 1:ncol(wide_df), batch = batch)

# Perform batch correction using harmonizR
invisible(harmonizR(wide_df, des_df, output_file = "in-house_batch_by_group"))
```

```{r,echo=FALSE}
#Read in data
cured_data_temp <- read_tsv('in-house_batch_by_group.tsv')
```

```{r, echo=FALSE}
# Using pivot_longer to transform to long format
long_cured_data <- cured_data_temp %>%
  pivot_longer(cols = -1, names_to = "Replicate Name", values_to = "Ratio To Standard")

#Separation of plate (batch) and wells
batch_by_group <- separate(long_cured_data, `...1`, into = c("Protein", "Peptide"), sep = "-") %>%
  separate(`Replicate Name`, into = c("Plate", "Well"), sep = "_")

batch_by_group$Plate <- as.numeric(sub(",", ".", batch_by_group$`Plate`))

# Extract last three characters from 'Well'
batch_by_group$Well <- substr(batch_by_group$Well, nchar(batch_by_group$Well) - 2, nchar(batch_by_group$Well))
```


```{r, median_centering, echo=FALSE}
# Median centering within each plate for each peptide
median_centered_data <- filtered_data %>%
  group_by(Plate, Peptide) %>%
  mutate(`Ratio To Standard` = `Ratio To Standard` - median(`Ratio To Standard`))
```


```{r, echo=FALSE}
# Selecting specific columns to create a new data frame
reduced_df <- median_centered_data[c("Plate","Well", "Protein", "Peptide", "Ratio To Standard")]

# Pivoting the data frame from long to wide format
reduced_df <- reduced_df %>%
group_by(Plate, Well, Protein, Peptide) %>%
  summarise(`Ratio To Standard` = max(`Ratio To Standard`))

wide_df_temp <- pivot_wider(reduced_df, names_from = c(Plate, Well), values_from = `Ratio To Standard`)
wide_df <- data.frame(wide_df_temp[, -(1:2), drop = FALSE])
col_names <- c(colnames(wide_df_temp))
colnames(wide_df) <- c(colnames(wide_df_temp))[-(1:2)]
row_names <- paste(wide_df_temp$Protein, wide_df_temp$Peptide, sep = "-")
rownames(wide_df) <- row_names

# Making a description dataframe where we take each well as a batch
batch_temp <- data.frame(colnames(wide_df))
batch <- separate(batch_temp, colnames.wide_df., into = c("Plate", "Well"), sep = "_")
batch <- c(as.integer(batch$Plate))

des_df <- data.frame(ID = colnames(wide_df), sample = 1:ncol(wide_df), batch = batch)

invisible(harmonizR(wide_df, des_df, output_file = "in-house_median_batch_by_plate"))
```

```{r, echo=FALSE}
#Read in data
cured_data_temp <- read_tsv('in-house_median_batch_by_plate.tsv')
```

```{r, echo=FALSE}
# Using pivot_longer to transform to long format

long_cured_data <- cured_data_temp %>%
  pivot_longer(cols = -1, names_to = "Replicate Name", values_to = "Ratio To Standard")

#Separation of plate (batch) and wells
mc_batch_by_plate <- separate(long_cured_data, `...1`, into = c("Protein", "Peptide"), sep = "-") %>%
  separate(`Replicate Name`, into = c("Plate", "Well"), sep = "_")

mc_batch_by_plate$Plate <- as.numeric(sub(",", ".", mc_batch_by_plate$`Plate`))

# Extract last three characters from 'Well'
mc_batch_by_plate$Well <- substr(mc_batch_by_plate$Well, nchar(mc_batch_by_plate$Well) - 2, nchar(mc_batch_by_plate$Well))
```

# Boxplot correction techniques comparison 

```{r, echo=FALSE, height=25, width=25}
# Create plots with specific dimensions and modifications

bp <- ggplot(batch_by_plate, aes(x = reorder(Plate, Well), y = log10(`Ratio To Standard`))) + 
  geom_boxplot(aes(color = Well)) +
  labs(title = "Batch corrected (by plate): ratio to standard well comparison",
       #x = "Plate-well",
       y = "Log10 of ratio to standard") +
  theme(legend.position = "none", 
        axis.title.x = element_blank(),
        plot.title = element_text(size = rel(0.8))) + 
  theme(axis.text.x = element_text(angle = 90)) + 
  ylim(-5,5) 

bg <- ggplot(batch_by_group, aes(x = reorder(Plate, Well), y = log10(`Ratio To Standard`))) + 
  geom_boxplot(aes(color = Well)) +
  labs(title = "Batch corrected (by group): ratio to standard per well",
       x = "Plate-well") +
  theme(legend.position = "none", 
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8))) +
  theme(axis.text.x = element_text(angle = 90)) + 
  ylim(-5,5)

mc <- ggplot(median_centered_data, aes(x = reorder(Plate, Well), y = log10(`Ratio To Standard`+1))) + 
  geom_boxplot(aes(color = Well)) +
  labs(title = "Median centered ratio to standard per well",
       x = "Plate-well",
       y = "Log10 of ratio to standard") +
  theme(legend.position = "none",  # Remove legend
        plot.title = element_text(size = rel(0.8))) +
  theme(axis.text.x = element_text(angle = 90)) +
  ylim(-5,5)

bmc <- ggplot(mc_batch_by_plate, aes(x = reorder(Plate, Well), y = log10(`Ratio To Standard`+1))) + 
  geom_boxplot(aes(color = Well)) +
  labs(title = "Median centered + batch corrected (by plate): ratio to standard per well",
       x = "Plate-well") +
  theme(legend.position = "none", 
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8))) +
  theme(axis.text.x = element_text(angle = 90)) +
  ylim(-5,5)

# Arrange plots without legends and with smaller titles
suppressWarnings({
  grid.arrange(bp, bg, mc, bmc, nrow = 2)
})
```

# APOA1 plot comparison


```{r, echo=FALSE, height=25, width=25}
#BY PLATE
#Set peptide colors 
peptide_colors <- c("VSFLSALEEYTK" = "#FF0000", "QGLLPVLESFK" = "#FF6666")

#APOA1 for batch correction by plate
apoa1_by_plate <- subset(batch_by_plate, Protein == "sp|P02647|APOA1_HUMAN")

#This together with the last line of the plot ensures that all plate numbers are visible
apoa1_by_plate$Plate <- factor(apoa1_by_plate$Plate)

#Visualize ratio to standard
apoa1.1 <- ggplot(apoa1_by_plate, aes(x = `Plate`, y = `Ratio To Standard` )) +
  geom_point(aes(color = `Peptide`)) +
  labs(title = "Batch corrected (by plate): APOA1 ratio to standard", y = "Ratio to standard") +
  scale_color_manual(values = peptide_colors) +
   theme(legend.position = "none", 
        axis.title.x = element_blank(),
        plot.title = element_text(size = rel(0.8))) +
  scale_x_discrete(labels = levels(apoa1_by_plate$Plate))+
  ylim(0, 45) 

#BY GROUP

#APOA1 for batch correction by group
apoa1_by_group <- subset(batch_by_group, Protein == "sp|P02647|APOA1_HUMAN")

#This together with the last line of the plot ensures that all plate numbers are visible
apoa1_by_group$Plate <- factor(apoa1_by_group$Plate)

#Visualize ratio to standard
apoa1.2 <- ggplot(apoa1_by_group, aes(x = `Plate`, y = `Ratio To Standard` )) +
  geom_point(aes(color = `Peptide`)) +
  labs(title = "Batch corrected (by group): APOA1 ratio to standard") +
  scale_color_manual(values = peptide_colors) +
   theme(legend.position = "none", 
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8)))   +
  scale_x_discrete(labels = levels(apoa1_by_group$Plate)) +
  ylim(0, 45) 

# MEDIAN CENTERED

#APOA1
apoa1_mc_data <- subset(median_centered_data, Protein == "sp|P02647|APOA1_HUMAN")

#This together with the last line of the plot ensures that all plate numbers are visible
apoa1_mc_data$Plate <- factor(apoa1_mc_data$Plate)

#Visualize ratio to standard
apoa1.3 <- ggplot(apoa1_mc_data, aes(x = Plate, y = `Ratio To Standard`)) +
  geom_point(aes(color = `Peptide`)) +
  labs(title = "Median centered APOA1 ratio to standard", x = "Plate", y = "Ratio to standard") +
  scale_color_manual(values = peptide_colors) +
  theme(legend.position = "none", 
        plot.title = element_text(size = rel(0.8)))  +
  scale_x_discrete(labels = levels(apoa1_mc_data$Plate))

#MEDIAN CENTERED BY PLATE

apoa1_mc_bp_data <- subset(mc_batch_by_plate, Protein == "sp|P02647|APOA1_HUMAN")

#This together with the last line of the plot ensures that all plate numbers are visible
apoa1_mc_bp_data$Plate <- factor(apoa1_mc_bp_data$Plate)

#Visualize ratio to standard
apoa1.4 <- ggplot(apoa1_mc_bp_data, aes(x = `Plate`, y = `Ratio To Standard` )) +
  geom_point(aes(color = `Peptide`)) +
  labs(title = "Median centered + batch corrected (by plate): APOA1 ratio to standard", x = "Plate") +
  scale_color_manual(values = peptide_colors) +
     theme(legend.position = "none", 
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8)))  +
  scale_x_discrete(labels = levels(apoa1_mc_bp_data$Plate))+
  ylim(-25, 20)

# Arrange plots without legends and with smaller titles
suppressWarnings({
  grid.arrange(apoa1.1, apoa1.2, apoa1.3, apoa1.4, nrow = 2)
})
```

# APOA1 smallest CV peptide, well pattern comparison

```{r, echo=FALSE, height=25, width=25}

# APOA1 peptide by plate
apoa1_pep_by_plate <- subset(batch_by_plate, Protein == "sp|P02647|APOA1_HUMAN") %>%
  subset(Peptide == "VSFLSALEEYTK")

well_colors <- c("A12" = "#990000", "B12" = "#FF3300", "C12" = "#FF9900")

apoa1_pep_by_plate$Plate <- factor(apoa1_pep_by_plate$Plate)
  
# Visualize ratio to standard for APOA1
apo_pep1 <- ggplot(apoa1_pep_by_plate, aes(x = Plate, y = `Ratio To Standard`, color = Well, group = Well)) +
  geom_point() +
  geom_line() +
  labs(title = "Batch corrected (by plate): APOA1-VSFLSALEEYTK ratio to standard", y = "Ratio to standard") +
  scale_color_manual(values = well_colors) +
   theme(legend.position = "none", 
        axis.title.x = element_blank(),
        plot.title = element_text(size = rel(0.8)))   +
  scale_x_discrete(labels = levels(apoa1_pep_by_plate$Plate))+
  ylim(-20, 40) 

# APOA1 peptide by group
apoa1_pep_by_group <- subset(batch_by_group, Protein == "sp|P02647|APOA1_HUMAN") %>%
  subset(Peptide == "VSFLSALEEYTK")

apoa1_pep_by_group$Plate <- factor(apoa1_pep_by_group$Plate)
  
# Visualize ratio to standard for APOA1
apo_pep2 <- ggplot(apoa1_pep_by_group, aes(x = Plate, y = `Ratio To Standard`, color = Well, group = Well)) +
  geom_point() +
  geom_line() +
  labs(title = "Batch corrected (by group): APOA1-VSFLSALEEYTK ratio to standard", x = "Plate", y = "Ratio to standard") +
  scale_color_manual(values = well_colors) +
   theme(legend.position = "none", 
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8))) +
  scale_x_discrete(labels = levels(apoa1_pep_by_group$Plate))+
  ylim(-20, 40) 


# MEDIAN CENTERED
apoa1_pep_mc <- subset(median_centered_data, Protein == "sp|P02647|APOA1_HUMAN") %>%
  subset(Peptide == "VSFLSALEEYTK")

apoa1_pep_mc$Plate <- factor(apoa1_pep_mc$Plate)
  
# Visualize ratio to standard for APOA1
apo_pep3 <- ggplot(apoa1_pep_mc, aes(x = Plate, y = `Ratio To Standard`, color = Well, group = Well)) +
  geom_point() +
  geom_line() +
  labs(title = "Median centered APOA1-VSFLSALEEYTK ratio to standard", x = "Plate", y = "Ratio to standard") +
  scale_color_manual(values = well_colors) +
   theme(legend.position = "none", 
        plot.title = element_text(size = rel(0.8))) +
  scale_x_discrete(labels = levels(apoa1_pep_mc$Plate))+
  ylim(-20, 40) 
  
# MEDIAN CENTERED, BATCH CORRECTED
apoa1_pep_mc_bp <- subset(mc_batch_by_plate, Protein == "sp|P02647|APOA1_HUMAN") %>%
  subset(Peptide == "VSFLSALEEYTK")

apoa1_pep_mc_bp$Plate <- factor(apoa1_pep_mc_bp$Plate)
  
# Visualize ratio to standard for APOA1
apo_pep4 <- ggplot(apoa1_pep_mc_bp, aes(x = Plate, y = `Ratio To Standard`, color = Well, group = Well)) +
  geom_point() +
  geom_line() +
  labs(title = "Median centered batch corrected (per plate) APOA1-VSFLSALEEYTK ratio to standard", x = "Plate", y = "Ratio to standard") +
  scale_color_manual(values = well_colors) +
   theme(legend.position = "none", 
        axis.title.y = element_blank(),
        plot.title = element_text(size = rel(0.8))) +
  scale_x_discrete(labels = levels(apoa1_pep_mc_bp$Plate))+
  ylim(-20, 40) 

# Arrange plots without legends and with smaller titles
suppressWarnings({
  grid.arrange(apo_pep1, apo_pep2, apo_pep3, apo_pep4, nrow = 2)
})

```

# References

Čuklina, J., Pedrioli, P. G., & Aebersold, R. (2020). Review of batch effects prevention, diagnostics, and correction approaches. *Mass spectrometry data analysis in proteomics*, 373-387.

Leek, J. T., Scharpf, R. B., Bravo, H. C., Simcha, D., Langmead, B., Johnson, W. E., ... & Irizarry, R. A. (2010). Tackling the widespread and critical impact of batch effects in high-throughput data. *Nature Reviews Genetics*, 11(10), 733-739.

Schlumbohm, S. (2022). *HarmonizR Software Instructions*. Available in: https://github.com/SimonSchlumbohm/HarmonizR/blob/main/HarmonizR_SOP.pdf [2023-12-05]

Voß, H., Schlumbohm, S., Barwikowski, P., Wurlitzer, M., Dottermusch, M., Neumann, P., ... & Krisp, C. (2022). HarmonizR enables data harmonization across independent proteomic datasets with appropriate handling of missing values. *Nature Communications*, 13(1), 3523.
